### PR TITLE
Add offset_of and owner_of

### PR DESCRIPTION
**Pull request by kghost**
`owner_of` function is able to eliminate several pointer in order to save memory.

### CODE REVIEW

File: src/lib/support/OwnerOf.h [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+#pragma once
+
+#include <cstdint>
+#include <cstddef>
+
+namespace chip {
+
+template<class ClassType, class MemeberType>
+static inline constexpr std::ptrdiff_t offset_of(const MemeberType ClassType::*member) {
```
Comment (Damian-Nordic) [inline src/lib/support/OwnerOf.h:27]: Can you use the standard https://en.cppreference.com/w/cpp/types/offsetof?

Comment (Damian-Nordic) [reply src/lib/support/OwnerOf.h:27]: Also, `constexpr` implies `inline`, so `constexpr` should be enough, here.

File: src/lib/support/OwnerOf.h [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    return reinterpret_cast<std::ptrdiff_t>(&(reinterpret_cast<ClassType*>(0)->*member));
+}
+
+template<class ClassType, class MemeberType>
+static inline constexpr ClassType* owner_of(MemeberType *ptr, const MemeberType ClassType::*member) {
```
Comment (Damian-Nordic) [inline src/lib/support/OwnerOf.h:32]: nit: `owner_of` -> `OwnerOf` to follow more common convention in CHIP?

File: src/lib/support/OwnerOf.h [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+ */
+
+#pragma once
+
+#include <cstdint>
+#include <cstddef>
+
+namespace chip {
+
+template<class ClassType, class MemeberType>
```
Comment (mspang) [inline src/lib/support/OwnerOf.h:26]: Did you mean `MemberType`?

Comment (kghost) [reply src/lib/support/OwnerOf.h:26]: Ohh, typo

**Review by Damian-Nordic — APPROVED** [review]

File: src/lib/support/OwnerOf.h [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+#pragma once
+
+#include <cstdint>
+#include <cstddef>
+
+namespace chip {
+
+template<class ClassType, class MemberType>
+constexpr ClassType* OwnerOf(MemberType *ptr, const MemberType ClassType::*member) {
+    return reinterpret_cast<ClassType*>(reinterpret_cast<uintptr_t>(ptr) - offsetof(ClassType, member));
```
Comment (bzbarsky-apple) [inline src/lib/support/OwnerOf.h:28]:
```suggestion
    static_assert(std::is_standard_layout<ClassType>::value, "offsetof does not make sense");
    return reinterpret_cast<ClassType*>(reinterpret_cast<uintptr_t>(ptr) - offsetof(ClassType, member));
```

File: src/lib/support/OwnerOf.h [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include <cstdint>
+#include <cstddef>
+
```
Comment (bzbarsky-apple) [inline src/lib/support/OwnerOf.h:23]:
```suggestion
#include <type_traits>

```

**Review by bzbarsky-apple — APPROVED** [review]

**Comment by bzbarsky-apple** [comment]
I guess there's the question whether this should be called `OwnerOf` or `ContainingObject`.....

File: src/lib/support/OwnerOf.h (hunk 4 above)
Comment (kghost) [reply src/lib/support/OwnerOf.h]: `is_standard_layout` is too restrictive, and is not necessary.

GCC will raise a compile error when `offsetof` is not applicable. Following code can not compile:

```
#include <cstddef>

struct A { int a; };
struct B : public virtual A   { int b; };

int main() {
    offsetof(A, A::a);
    offsetof(B, B::a);
    return 0;
}
```

```
% gcc -std=c++11 offsetof.cpp -lstdc++
In file included from /usr/include/c++/10/cstddef:50,
                 from offsetof.cpp:1:
offsetof.cpp: In function ‘int main()’:
offsetof.cpp:8:20: error: invalid access to non-static data member ‘A::a’ in virtual base of NULL object
    8 |     offsetof(B, B::a);
      |                    ^
```

Comment (bukepo) [reply src/lib/support/OwnerOf.h]: It's undefined behavior in C++11. Is CHIP using C++17?

**Comment by msandstedt** [comment]
With this feature you can replace a function signature that looks like `fn(a, b)` with `fn(b)`.  So yes, you save ~one word of stack space in this call, likely trading it for code space for the template.  Likewise, if you needed to store both pointers, now you can potentially only store one and save there.

These seem like very modest improvements in memory footprint, and again would likely have a corresponding modest cost in code space.

My question is: is this worth it and needed?  This seems like a rather exotic c++ feature to leverage.  Will this lead to good style?  Do we have a style guide that can tell us whether we collectively believe this is a good thing to do?

**Review by yufengwangca — APPROVED** [review]

**Comment by yufengwangca** [comment]
@kghost Could you address the remaining items? and close those already addressed

File: src/lib/support/OwnerOf.h (hunk 4 above)
Comment (kghost) [reply src/lib/support/OwnerOf.h]: In practical, we will get an error when it is undefined behavior. And in practical, even it is not standard layer, it is still well defined by implementation (GCC or LLVM). That's why I didn't add the check.

Comment (bzbarsky-apple) [reply src/lib/support/OwnerOf.h]: > In practical, we will get an error when it is undefined behavior.

What guarantees that?

Why is standard layout too restrictive?

**Pull request merged**